dashboard/config/custom_session_store/migrate_cookies_to_redis_store.rb

```diff
@@ -0,0 +1,59 @@
+require 'rack/session/abstract/id'
+require 'redis-actionpack'
+
+module ActionDispatch
+  module Session
+    # Temporary helper module for the transition from cookie-based to
+    # redis-based session storage. If a request contains actual session data
+    # rather than just a session identifier, seamlessly write the session data
+    # to redis. If a session has already been migrated, do nothing.
+    #
+    # TODO infra: once all user sessions have been migrated, remove this helper
+    # and start using a non-customized implementation of redis-based session
+    # storage.
+    module MigrateCookiesStore
+      # Ultimately loads session data from the database, after first checking
+      # for session data in the cookie and persisting it to the database if we
+      # find any.
+      def load_session(request)
+        # migrate_session_data will return the extracted session id if there
+        # was one; otherwise, we fetch the session id from the request object.
+        session_id = migrate_session_data(request) || current_session_id(request)
+        session_id, session = find_session(request, session_id)
+        [session_id, session || {}]
+      end
+
+      # If we already have existing session data in the cookie, this helper
+      # method will clear the cookie, write that session data to the database,
+      # and return the session id it extracted from the cookie.
+      #
+      # If the cookie does not contain any existing session data, do nothing
+      # and return nothing.
+      private def migrate_session_data(request)
+        stale_session_check! do
+          session_data = unpacked_cookie_data(request)
+          if session_data.is_a?(Hash) && !session_data.empty?
+            session_id = Rack::Session::SessionId.new(session_data["session_id"])
+            request.set_header("action_dispatch.request.unsigned_session_cookie", {})
+            write_session(request, session_id, session_data.except("session_id"), request.session_options)
```

Comment:
Between these two lines of code, the session only exists in memory, is there anything bad that happens if we clear the cookie and then fail to write the session? Or is it fine if we throw a 500 with the cookie cleared?

Reply:
Oh, good catch! I'm not entirely sure how the 500 error response interacts with manually-set headers like this, but my expectation is that we would indeed lose the cookie data in that case. I'll swap these lines, and add an error handler just in case.